### ISSUE TITLE
How to use access level?

### ISSUE THREAD

**Issue by ferologics** [scribd/Weaver#31]
The framework API for end user shows all of the generated Containers/Resolvers if `public` access level is used in container.

Desired behaviour:
```swift
public class Foo { declare dependencies... }
```
Omit the`public` access level declaration in `Weaver.Foo.swift`.
```swift
final class FooDependencyContainer: DependencyContainer { ... }
protocol FooDependencyResolver { ... }
extension FooDependencyContainer: FooDependencyResolver { ... }
```

I see that this has been implemented in [#2](https://github.com/scribd/Weaver/pull/2).

This configuration annotation should be added as a new `ConfigurationAttribute` called `accessLevel`.
```swift
public class Foo {
  // weaver: cloudEndpoint = CloudEndpoint
  // weaver: cloudEndpoint.accessLevel = .internal
}
```

This unexpected behaviour should be clearly documented, since I am blocked from shipping anything to production until proper access control is established. 

Default access level should be set to `.internal` per current implementation.

**Comment by trupin** [comment]
Hi @ferologics,

This is a known limitation which is more due to the language itself, which I could indeed document by providing an example.

You can't provide an `internal` dependency via a `public` resolver. This comes from the fact that an injectable type resolves its dependencies from a unique `protocol`, which needs to be `public`. Making this `protocol` public requires all the injected types to be `public` as well. 

This behavior seems right to me since its matching the same requirements than what a regular `public` initializer would have. 

For example, in the code below, `CloudEndPoint` has to be `public` if `Foo` wants to have any chance to get initialized from outside the `internal` scope.

```swift
public final class Foo {
    public init(couldEntPoint: CloudEndPoint) {
        ...
    }
}
```

I would probably be able to help more with more input on what you're trying to do exactly.

If you want to expose a type's API but, not its initializer, I suggest exposing a `protocol` which your internal type would implement instead. Like the following

```swift
public protocol FooProtocol {
    func helloWorld()
}
final class Foo: FooProtocol {
    init(couldEntPoint: CloudEndPoint) {
        ...
    }

    func helloWorld() { ... }
}
```

**Comment by trupin** [comment]
Also, I haven't used Weaver to do dependency injection in a framework so far, so it's possible that I don't see all the limitations of this.

I'll add this use case to the sample in order to demonstrate how to publicly expose dependencies, and see if there are improvements that are needed.

**Comment by ferologics** [comment]
I will try your suggestion, though I am not sure if it's what I'm looking for. Either way, many thanks.

The goal is to package a framework without exposing any of the generated swift dependencies that follow from the public container dependencies.

```swift
public class API { declare dependencies and expose API functionality... }
```

The end user should not have a view in to the internally generated dependencies.

```swift
public final class APIDependencyContainer: DependencyContainer { ... }
public protocol APIDependencyResolver { ... }
public extension APIDependencyContainer: FooDependencyResolver { ... }
```

As I see it, the generated classes / protocols shouldn't be packaged and made visible in the public interface of the framework API, somehow they should be internal to the implementation of the framework.

**Comment by trupin** [comment]
Ok, I see better your point now.

You are entirely right, we'd need an additional `accessLevel` attribute set to `internal` by default for in-place registrations like `// weaver: foo = Foo <- FooProtocol`, and to `public` by default for references like `// weaver: foo <- FooProtocol`.

So, adding the attribute is the easy part. The challenging part will be to generate the DI Container in a way that allows to only expose only the `public` dependencies, but not the `internal` dependencies. I don't want to get away from the generated `FooDependencyResolver` protocol because it's very helpful for testing.

Maybe a solution could be to have two resolver protocols, one `internal`, the other `public` like so:

```swift
public final class APIDependencyContainer: DependencyContainer { ... }
public protocol APIDependencyResolver { ... }
protocol APIInternalDependencyResolver: APIDependencyResolver { ... }
public extension APIDependencyContainer: APIInternalDependencyResolver { ... }
```

Two initializers would then need to be implemented, one `internal`, the other `public` like so:

```swift
public final class API {
    init(_ internalDependencies: APIInternalDependencyResolver) { ... }
    public init(_ dependencies: APIDependencyResolver) { ... }
}
```

I'll look into that tonight and I'll keep you posted on my progress. But if you really need to ship your code asap, I guess you could write this code that Weaver is supposed to generate manually, and then remove it as soon as the issue is resolved

**Comment by trupin** [comment]
I think my first idea is overly complicated, and an internal protocol can't inherit from a public protocol anyway.

I worked on an example of generated code and here is what I came up with:

Let's say we have the following class:
```swift
public final class API {
    // weaver: logger = Logger
    // weaver: urlSession <- URLSession
}
```

The generated code would be the following:
```swift
// MARK: - MovieAPI

class MovieAPIDependencyContainer: DependencyContainer {
    init(parent: DependencyContainer? = nil) {
        super.init(parent)
    }
    override func registerDependencies(in store: DependencyStore) {
        store.register(Logger.self, scope: .transient, name: "logger") { _ in
            return Logger()
        }
    }
}
protocol MovieAPIDependencyResolver {
    var urlSession: URLSession { get }
    var logger: Logger { get }
}
extension MovieAPIDependencyContainer: MovieAPIDependencyResolver {
    var urlSession: URLSession {
        return resolve(URLSession.self, name: "urlSession")
    }
    var logger: Logger {
        return resolve(Logger.self, name: "logger")
    }
}
extension MovieAPI {
    public static func makeMovieAPI(injecting parentDependencies: DependencyContainer) -> MovieAPI {
        let dependencies = MovieAPIDependencyContainer(parent: parentDependencies)
        return MovieAPI(injecting: dependencies)
    }
}
protocol MovieAPIDependencyInjectable {
    init(injecting dependencies: MovieAPIDependencyResolver)
}
extension MovieAPI: MovieAPIDependencyInjectable {}

// MARK: - MovieAPI Shim

final class MovieAPIShimDependencyContainer {
    let urlSession: URLSession

    private let internalDependencies = MovieAPIDependencyContainer()
    
    init(urlSession: URLSession) {
        self.urlSession = urlSession
    }
}
extension MovieAPIShimDependencyContainer: MovieAPIDependencyResolver {
    var logger: Logger {
        return internalDependencies.resolve(Logger.self, name: "logger")
    }
}
extension MovieAPI {
    public convenience init(urlSession: URLSession) {
        self.init(injecting: MovieAPIShimDependencyContainer(urlSession: urlSession))
    }
}
```

This way, `MovieAPI` can be used inside the framework as a normal dependency, but also can be used from outside the framework, and be initialized with the `public` convenience initializer, hiding completely Weaver.

Also, `Logger` remains internal, while `URLSession` is public and can be injected from an external call site.

**Comment by ferologics** [comment]
Wow, this actually looks doable. Great job! 💪

Do you have this somewhere on branch? I'd love to test this in my project ASAP.
Hit me up w/ the PR for review if that would speed things up ✌️

**Comment by trupin** [comment]
Hey @ferologics, 

I have been pretty busy with other matters lately, but I have been able to try Weaver on a framework of mine. Unfortunately, I caught several blocking issues. I I'll have time to work on those next week.

**Comment by ferologics** [comment]
Hey @trupin, no worries. 

I have forsaken Weaver in my project for now but I hope to integrate it sooner or later.
Better still, as I genuinely find the project interesting, I can allocate some time if needed.

**Comment by trupin** [comment]
Closing this ticket since it has been resolved by #32.

@ferologics, your feedback on v0.9.10 is welcome!
